### PR TITLE
add robust scope names to RAM profiler for torch>=2.0

### LINKED ISSUE #393: RAM profile fails to capture YOLO module names
Module names under the `scope` column of the RAM profile table do not work for YOLO models due to the failed capture of the Sequential module. In the case where the Sequential module is iterated over with a for loop, the module was not captured, because its forward function was not called.

From torch 1.13, scope name tracing is natively supported with `torch.onnx.utils._setup_trace_module_map`, so I added that option to our trace handling for torch versions that support it. This change handles the Sequential module edge case and resolves the issue with the YOLO module name tracing. However, this fix only resolves the case where users are using `torch>=1.13` (most users)

### PR DESCRIPTION
**Pull request by alexh-deeplite**
Resolves #393:

Module names under the `scope` column of the RAM profile table do not work for YOLO models due to the failed capture of the Sequential module. In the case where the Sequential module is iterated over with a for loop, the module was not captured, because its forward function was not called.

From torch 1.13, scope name tracing is natively supported with `torch.onnx.utils._setup_trace_module_map`, so I added that option to our trace handling for torch versions that support it. This change handles the Sequential module edge case and resolves the issue with the YOLO module name tracing. This fix only resolves the case where users are using `torch>=2.0` (most users), but we are no longer actively supporting torch<2.0 anyways so I feel this is acceptable

### CODE REVIEW

**Comment by alexh-deeplite** [comment]
will wait for setuptools bug to be fixed by merging #396  first

File: deeplite_torch_zoo/utils/profiler/utils/trace.py [hunk 1]
```diff
@@ -80,8 +133,13 @@ def trace(model, args=(), kwargs=None):
         'Please use positional arguments instead!'
     )
 
-    with warnings.catch_warnings(record=True), ScopeNameContextManager():
+    # initiliaze module scope caching, supported as of torch 1.13
```
Comment (matteog11) [inline deeplite_torch_zoo/utils/profiler/utils/trace.py:136]: Typo

File: deeplite_torch_zoo/utils/profiler/utils/trace.py [hunk 2]
```diff
@@ -61,6 +62,58 @@ def _get_tracing_name(module, tracing_state):
...
+        if len(type_name) == 1:
+            module_names.append(type_name[0])
+
+        module_type, module_name = type_name[0], type_name[-1]
+
+        module_name_split = module_name.split('.')
+        if len(module_name_split) == 1:
+            module_names.append(module_name)
+        else:
+            # we are in child of sequential, so module name includes parent name
```
Comment (matteog11) [inline deeplite_torch_zoo/utils/profiler/utils/trace.py:91]: @alexh-deeplite Could you provide a clearer and more structured explanation comment on handling nested modules within Sequential (maybe in the header)?

File: deeplite_torch_zoo/utils/profiler/utils/trace.py [hunk 3]
```diff
@@ -61,6 +62,58 @@ def _get_tracing_name(module, tracing_state):
...
+            module_names_idx = len(module_names) - 1
+            split_idx = len(module_name_split) - 2
+            last_saved_name = module_names[module_names_idx]
+            # from parent of module, check if parent in the stack, if not add it to names to add
+            for split_idx in range(len(module_name_split) - 2, -1, -1):
+                parent_name = module_name_split[split_idx]
+                if parent_name == last_saved_name:
+                    break
+                names_to_add.append(parent_name)
+                types_to_add.append('__hidden_sequential__')
```
Comment (matteog11) [inline deeplite_torch_zoo/utils/profiler/utils/trace.py:107]: Could you add more details on the use of __hidden_sequential__?

File: deeplite_torch_zoo/utils/profiler/utils/trace.py (hunk 2 above)
Comment (alexh-deeplite) [reply deeplite_torch_zoo/utils/profiler/utils/trace.py:91]: I'll add a docstring to explain some of these steps in more detail!

File: deeplite_torch_zoo/utils/profiler/utils/trace.py (hunk 3 above)
Comment (alexh-deeplite) [reply deeplite_torch_zoo/utils/profiler/utils/trace.py]: added comment (line is for debugging purposes)

**Comment by alexh-deeplite** [comment]
Linking the related pytorch issue to share our workaround [filter_torch_scope_name](https://github.com/Deeplite/deeplite-torch-zoo/blob/7d1aca8e210057db40c844a2d456a4a1cdaf3a3c/deeplite_torch_zoo/utils/profiler/utils/trace.py#L64) 

https://github.com/pytorch/pytorch/issues/90439

**Pull request merged**